### DISCUSSION TITLE
Merge C libraries into spork

### DISCUSSION THREAD

**Discussion by bakpakin** [janet-lang/janet#982] [Ideas]
Currently, [spork](https://github.com/janet-lang/spork) is a utility library for Janet, but is limited to pure Janet only. However, by merging things like the `json` package (which is written in C) into spork, I think spork itself can be more useful and easier to maintain. Thoughts? I can see some people might like to keep things like JSON separate, but I think it is better to bundle things together for ease of development.

**Comment by pepe** [comment]
I started in the no C camp. But now I see it as a very fine idea.

**Reply by pepe** [reply]
But possibly only to those without any dependencies? So no jhydro, miniz and friends.

**Reply by bakpakin** [reply]
Yeah, not sure where to draw a line. My current thought is that the libraries shouldn't be configurable or limited to some systems, need to be MIT licensed, and preferably not too large. But something like json, typed arrays, etc. could be put in for centralization. It's easier for me to maintain a single contrib library compared to many small libraries, and there is no runtime cost either. The only cost would be trying to install spork on a system without a C compiler, but to me that just signals an invalid jpm install.

**Reply by pepe** [reply]
I would start with JSON and typed array. I guess we will see if there are any issues pretty soon. I would like to merge all my libraries from a Good Place into one big package too.

**Comment by sogaiu** [comment]
Along the lines of "without any dependencies", how about not growing any additional dependencies (on things already in spork) if merged?  I think that would make the json bits still easily usable a la carte without being influenced by other parts of spork.

**Reply by bakpakin** [reply]
Yeah, modules in spork would be as much as possible separate - JSON is never going to depend on anything else in spork since it doesn't need to

**Reply by pepe** [reply]
With one exception being a test?

**Comment by uvtc** [comment]
One thing that attracts me to Janet is that it's a native program implemented in C. So having native modules in spork ("contrib") is a good thing in my book, as long as it's clear what native libraries (better yet, Debian packages) that I need to have already installed as prerequisites. (Or, when I do `jpm install spork`, if missing dependencies, it would tell me, "Please install prerequisite ${native-lib} in order to install spork".)

**Comment by uvtc** [comment]
Another possible way forward: suppose in the future there's some good native modules for Janet written in something other than C (like, for example, [Hare](https://harelang.org/), [Zig](https://ziglang.org/), [Rust](https://www.rust-lang.org/), etc.) ... modules that you'd like to be part of spork. I wonder if it would then make sense to have separate projects, one for each native implementation language:

spork-c\
spork-hare\
spork-zig\
spork-rust
